x/wasm/wasm_snapshotter.go

```diff
@@ -0,0 +1,150 @@
+)
+
+/*
+API to implement:
+
+// Snapshotter is something that can create and restore snapshots, consisting of streamed binary
+// chunks - all of which must be read from the channel and closed. If an unsupported format is
+// given, it must return ErrUnknownFormat (possibly wrapped with fmt.Errorf).
+type Snapshotter interface {
+	// Snapshot writes snapshot items into the protobuf writer.
+	Snapshot(height uint64, protoWriter protoio.Writer) error
+
+	// Restore restores a state snapshot from the protobuf items read from the reader.
+	// If the ready channel is non-nil, it returns a ready signal (by being closed) once the
+	// restorer is ready to accept chunks.
+	Restore(height uint64, format uint32, protoReader protoio.Reader) (SnapshotItem, error)
+}
+
+// ExtensionSnapshotter is an extension Snapshotter that is appended to the snapshot stream.
+// ExtensionSnapshotter has an unique name and manages it's own internal formats.
+type ExtensionSnapshotter interface {
+	Snapshotter
+
+	// SnapshotName returns the name of snapshotter, it should be unique in the manager.
+	SnapshotName() string
+
+	// SnapshotFormat returns the default format the extension snapshotter use to encode the
+	// payloads when taking a snapshot.
+	// It's defined within the extension, different from the global format for the whole state-sync snapshot.
+	SnapshotFormat() uint32
+
+	// SupportedFormats returns a list of formats it can restore from.
+	SupportedFormats() []uint32
+}
+*/
+
+type WasmSnapshotter struct {
+	wasmDirectory string
+}
+
+func NewWasmSnapshotter(wasmDirectory string) *WasmSnapshotter {
+	return &WasmSnapshotter{
+		wasmDirectory,
+	}
+}
+
+func (ws *WasmSnapshotter) SnapshotName() string {
+	return "WASM Files Snapshot"
+}
+
+func (ws *WasmSnapshotter) SnapshotFormat() uint32 {
+	return 1
+}
+
+func (ws *WasmSnapshotter) SupportedFormats() []uint32 {
+	return []uint32{1}
+}
+
+var wasmFileNameRegex = regexp.MustCompile(`^[a-f0-9]{64}$`)
+
+func (ws *WasmSnapshotter) Snapshot(height uint64, protoWriter protoio.Writer) error {
+	wasmFiles, err := ioutil.ReadDir(ws.wasmDirectory)
+	if err != nil {
+		return err
+	}
+
+	// In case snapshotting needs to be deterministic
+	sort.SliceStable(wasmFiles, func(i, j int) bool {
+		return strings.Compare(wasmFiles[i].Name(), wasmFiles[j].Name()) < 0
+	})
+
+	for _, wasmFile := range wasmFiles {
+		if !wasmFileNameRegex.MatchString(wasmFile.Name()) {
+			continue
+		}
+
+		wasmFilePath := filepath.Join(ws.wasmDirectory, wasmFile.Name())
+		wasmBytes, err := ioutil.ReadFile(wasmFilePath)
+		if err != nil {
+			return err
+		}
+
+		// snapshotItem is 64 bytes of the file name, then the actual WASM bytes
+		snapshotItem := append([]byte(wasmFile.Name()), wasmBytes...)
+
+		snapshot.WriteExtensionItem(protoWriter, snapshotItem)
+	}
+
+	return nil
+}
+
+func (ws *WasmSnapshotter) Restore(
+	height uint64, format uint32, protoReader protoio.Reader,
+) (snapshot.SnapshotItem, error) {
```

Comment:
This seems to read the entire data and return nothing ever. Why does it return `(snapshot.SnapshotItem, error)` rather than `error`? (eg. Is there something I/we are missing?)

Reply:
I know this is not your design, but wonder why it is there

Reply:
Yeah, I took this from the sdk tests. Not sure if the return value goes anywhere here, but I can confirm that this works without any issue (tested manually on Secret). Also might be irrelevant for this case as we're dealing with files and not IAVL.

Reply:
All this logic is in a loop where it seems like its the extension responsibility to handle the next `SnapshotItem` to the next one, which doesn't make sense because a plugin can break another one without knowing.

See https://github.com/cosmos/cosmos-sdk/blob/6f070623741fe0d6851d79ada41e6e2b1c67e236/snapshots/manager.go#L333

Reply:
> My guess is that is possible that in the protobuf stream there might be more items and if the format is not the one that wasmd can understand then it should return it.

Yes, very good idea. It seems like we should not error on some unknown extension, but just return it, as there may in the future be multiple extensions than need to cooperate somehow. Not sure how this all works together.

Maybe @yihuang could clarify how we should handle this properly

Reply:
> Yes, very good idea. It seems like we should not error on some unknown extension, but just return it, as there may in the future be multiple extensions than need to cooperate somehow. Not sure how this all works together.

exactly, we should return the item if it's unknown to us.

Reply:
> which doesn't make sense because a plugin can break another one without knowing.

yeah, that's possible unfortunately, maybe we can create some helpers at module side without changing the cosmos-sdk api.

Reply:
ah, I got it... we *either* end on EOF, or when we hit something where `GetExtensionPayload()` returns nil (another type we don't expect)

Reply:
> > which doesn't make sense because a plugin can break another one without knowing.
> 
> yeah, that's possible unfortunately, maybe we can create some helpers at module side to retain the last item automatically, without changing the api in cosmos-sdk.

```

// ExtensionPayloadReader only reads the payloads of current extension.
type ExtensionPayloadReader struct {
	protoReader protoio.Reader
	LastItem    types.SnapshotItem
}

func NewExtensionPayloadReader(protoReader protoio.Reader) *ExtensionPayloadReader {
	return &ExtensionPayloadReader{
		protoReader: protoReader,
		LastItem:    types.SnapshotItem{},
	}
}

// ReadPayload reads a payload item from the stream.
func (epr *ExtensionPayloadReader) ReadPayload() ([]byte, error) {
	epr.LastItem = types.SnapshotItem{}
	if err := epr.protoReader.ReadMsg(&epr.LastItem); err != nil {
		return nil, err
	}
	payload := epr.LastItem.GetExtensionPayload()
	if payload == nil {
		// end of payload stream of current extension
		return nil, io.EOF
	}
	return payload.Payload, nil
}
```

I think of sth like this could help, the manager pass the `ExtensionPayloadReader` extension's `Restore` method, but it's api breaking, so maybe change this in next version.